src/python/pants/engine/target.py

```diff
@@ -0,0 +1,194 @@
+from typing import Any, ClassVar, Dict, Iterable, Optional, Tuple, Type, TypeVar, cast
+
+from pants.engine.rules import UnionMembership
+from pants.util.memo import memoized_property
+from pants.util.meta import frozen_after_init
+
+
+@dataclass(frozen=True)
+class Field(ABC):
+    alias: ClassVar[str]
+    raw_value: Optional[Any]  # None indicates that the field was not explicitly defined
+
+    def __repr__(self) -> str:
+        return f"{self.__class__}(alias={repr(self.alias)}, raw_value={self.raw_value})"
+
+
+class PrimitiveField(Field, metaclass=ABCMeta):
+    """A Field that does not need the engine in order to be hydrated.
+
+    This should be subclassed by the majority of fields.
+    """
+
+    def __str__(self) -> str:
+        return f"{self.alias}={self.value}"
+
+    @memoized_property
+    @abstractmethod
+    def value(self) -> Any:
+        """Get the field's value.
+
+        The value will possibly be first hydrated and/or validated, such as using a default value
+        if the field was not defined or ensuring that an int value is positive.
+
+        This property is memoized because hydration and validation can often be costly. This
+        hydration is lazy, i.e. it will only happen when a downstream rule explicitly requests this
+        field.
+        """
+
+
+class AsyncField(Field, metaclass=ABCMeta):
+    """A field that needs the engine in order to be hydrated.
+
+    You should create a corresponding Result class and define a rule to go from this AsyncField to
+    the Result. For example:
+
+        class Sources(AsyncField):
+            alias: ClassVar = "sources"
+            raw_value: Optional[List[str]]
+
+
+        @dataclass(frozen=True)
+        class SourcesResult:
+            snapshot: Snapshot
+
+
+        @rule
+        def hydrate_sources(sources: Sources) -> SourcesResult:
+            sources.validate_pre_hydration()
+            result = await Get[Snapshot](PathGlobs(sources.raw_value))
+            sources.validate_post_hydration(result)
+            return SourcesResult(result)
+
+
+        def rules():
+            return [hydrate_sources]
+
+    Then, call sites can `await Get` if they need to hydrate the field:
+
+        sources = await Get[SourcesResult](Sources, my_tgt.get(Sources))
+    """
+
+    def validate_pre_hydration(self) -> None:
+        """Any validation that can be done on the original `raw_value`.
+
+        It is cheaper to do any possible validation here, rather than in `validate_post_hydration`,
+        because we can short-circuit if an invariant is violated before incurring the expense of
+        hydration.
+        """
+
+    def validate_post_hydration(self, result: Any) -> None:
+        """Any validation that must be done after hydration by inspecting the result of that
+        hydration.
+
+        For example, a `PythonSources` field may validate that all hydrated source files end in
+        `.py`.
+        """
+
+
+class PluginField:
```

Comment:
Why does this need to exist separate from `Field`? And if it needs to exist, why isn't it a subclass of `Field`?

Reply:
This is a sentinel type. We don't expect it to correspond to any actual `Field` in a `BUILD` file. It solely exists so that plugin authors can say `UnionRule(PythonLibraryField, TypeChecked)`.

There are other possibilities:
1. Name this something better to reflect that it's a sentinel and not a normal `Field`.
2. Instead of `UnionRule(PythonLibraryField, TypeChecked)`, have `UionRule(PythonLibrary, TypeChecked)`. That is, expect `UnionRule(Target, Field)` instead of `UnionRule(PluginField, Field)`.
   * This reduces boilerplate pretty substantially.
   * Conceptually, it doesn't make as much sense to me as `PluginField`. For example, now you'd have to mark `PythonLibrary` with `@union`. What does that mean? That suggests to me that you can substitute `PythonLibrary` with other union members, but that's not what we're actually saying.
   * Maybe we can avoid having to mark each target with `@union`..? Maybe that can get done under-the-hood? We never actually use `await Get[A](UnionBase, member)` with `PluginField`. We solely (ab)use `UnionMembership` in this constructor, so maybe it doesn't even matter if we mark `@union` anywhere?